File: mobile-app/lib/src/core/models/command.dart
/// An abstract base class for commands sent over the WebSocket.
///
/// This class has a concrete factory method that takes a JSON object and a box ID
/// and returns an instance of a concrete command class based on the command type
/// in the JSON object.
abstract class Command {
  final int boxId;

  Command(this.boxId);

  factory Command.fromJson(Map<String, dynamic> json, int boxId) {
    String? commandType = json['commandType'];
    String? uniqueId = json['uniqueId'];
    switch (commandType) {
      case 'Connection':
        return Connection(boxId, uniqueId!);
      case 'StartGameCommand':
        return StartGameCommand(boxId);
      case 'StopGameCommand':
        return StopGameCommand(boxId);
      case 'VoteCommand':
        return VoteCommand(boxId, uniqueId!, json['message']);
      case 'ConnectRemote':
        return ConnectRemote(boxId, uniqueId!);
      case 'DisconnectRemote':
        return DisconnectRemote(boxId, uniqueId!);
      case 'StatusCommand':
        return StatusCommand(boxId, json['status'], json['message']);
      case 'StickExploded':
        return StickExploded(boxId);
      case 'VoteResult':
        return VoteResult(boxId);
      case 'AnnecdotTeller':
        return AnnecdotTeller(boxId);
      case 'RetrieveStateCommand':
        return RetrieveStateCommand(boxId);
      case 'GameModeChanged':
        return GameModeChanged(boxId, json["message"]);
      case 'ClientDisconnected':
        return ClientDisconnected(boxId, json["senderUniqueId"]);
      case 'SubjectChanged':
        return SubjectChanged(boxId, json["message"]);
      case 'StickScanned':
        return StickScanned(boxId, uniqueId!);
      case 'PlayStickExploded':
        return PlayStickExploded(boxId);
      default:
        throw UnsupportedError('Unknown command type: $commandType');
    }
  }
}

class Connection extends Command {
  final String uniqueId;
  Connection(super.boxId, this.uniqueId);
}

class StartGameCommand extends Command {
  StartGameCommand(super.boxId);
}

class StartRoundCommand extends Command {
  StartRoundCommand(super.boxId);
}

class StopGameCommand extends Command {
  StopGameCommand(super.boxId);
}

class VoteCommand extends Command {
  final String uniqueId;
  final String vote;

  VoteCommand(super.boxId, this.uniqueId, this.vote);
}

class ConnectRemote extends Command {
  final String uniqueId;

  ConnectRemote(super.boxId, this.uniqueId);
}

class DisconnectRemote extends Command {
  final String uniqueId;

  DisconnectRemote(super.boxId, this.uniqueId);
}

class StatusCommand extends Command {
  final String status;
  final String message;
  StatusCommand(super.boxId, this.status, this.message);
}

class StickExploded extends Command {
  StickExploded(super.boxId);
}

class VoteResult extends Command {
  VoteResult(super.boxId);
}

class AnnecdotTeller extends Command {
  AnnecdotTeller(super.boxId);
}

class RetrieveStateCommand extends Command {
  RetrieveStateCommand(super.boxId);
}

class GameModeChanged extends Command {
  final String gameMode;

  GameModeChanged(super.boxId, this.gameMode);
}

class ClientDisconnected extends Command {
  final String disconnectedUserId;

  ClientDisconnected(super.boxId, this.disconnectedUserId);
}

class SubjectChanged extends Command {
  final String subject;

  SubjectChanged(super.boxId, this.subject);
}

class StickScanned extends Command {
  final String uniqueId;

  StickScanned(super.boxId, this.uniqueId);
}

class PlayStickExploded extends Command {
  PlayStickExploded(super.boxId);
}

File: server/src/main/scala/com/anecdot/Command.scala
package com.anecdot

import akka.actor.typed.ActorRef
import akka.http.scaladsl.model.ws.TextMessage

sealed trait Command {
  def boxId: Int
}

case class StartGameCommand(boxId: Int, uniqueId: String) extends Command
case class StartRoundCommand(boxId: Int, uniqueId: String) extends Command
case class StopGameCommand(boxId: Int, uniqueId: String) extends Command
case class VoteCommand(boxId: Int, vote: String, uniqueId: String) extends Command
case class ConnectRemote(boxId: Int, uniqueId: String, username: String) extends Command
case class DisconnectRemote(boxId: Int, uniqueId: String) extends Command
case class ConnectBox(boxId: Int, uniqueId: String) extends Command
case class StartVoting(boxId: Int, uniqueId: String) extends Command
case class StickExploded(boxId: Int) extends Command
case class ScannedStickCommand(boxId: Int, uniqueId: String, exploded : Boolean) extends Command

case class ExplodedAnimationPlayed(boxId: Int) extends Command

case class StopRoundCommand(boxId: Int) extends Command
case class SetGameModeCommand(boxId: Int, uniqueId: String, gameMode: String) extends Command
case class IdleGameCommand(boxId: Int) extends Command

case class RetrieveStateCommand(boxId: Int, uniqueId: String) extends Command

enum DisconnectReason:
  case ClientDisconnected, LostConnection

case class ClientDisconnected(boxId: Int, uniqueId: String, reason: DisconnectReason ) extends Command
case class HeartbeatClient(boxId: Int, uniqueId: String) extends Command
case class CommandResponse(
    uniqueId: String,
    commandType: String,
    status: String,
    message: Option[String] = None,
    senderUniqueId: Option[String] = None,
    scores: Option[Map[String, Int]] = None
)

case class GameStateSnapshot(
  players: Map[String, Map[String, String]], // If the inner map contains only string values
  playerScores: Map[String, Int],
  stickExploded: Boolean,
  annecdotTellerId: String,
  state: String
)

case class VoiceFlow(boxId: Int, uniqueId: String, payload: Option[String] ) extends Command

case class VoteSubmittedNotification(vote: String, uniqueId: String) extends Command {
  def boxId: Int = -1
}

case class GameStateChangedNotification(newState: States, uniqueId: String) extends Command {
  def boxId: Int = -1
}

case class AddClient(boxId: Int, uniqueId: String, ref: ActorRef[TextMessage]) extends Command
case class CheckInactiveClients() extends Command {
  // This is a special command that is not associated with a boxId because it is not sent from a client
  def boxId: Int = -1
}
